gourmet/gtk/recipe.py

```diff
@@ -0,0 +1,787 @@
+}
+
+
+class _ImageData:
+
+    def __init__(self, data=None):
+        self.data = data
+        self.edited = False
+
+
+
+class _IngredientGroup:
+
+    def __init__(self, iter=None):
+        self.iter = iter
+
+
+
+class _IngredientList(_Gtk.TreeStore):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+
+    def do_row_drop_possible(self, dest_path, data):
+        depth = dest_path.get_depth()
+        if 1 == depth:
+            return True
+        if 2 == depth:
+            parent = dest_path.copy()
+            parent.up()
+            iter = self.get_iter(parent)
+            # Only allow drops onto an ingredient group
+            if not isinstance(self.get_value(iter, 4), _IngredientGroup):
+                return False
+
+            assert 'GTK_TREE_MODEL_ROW' == data.get_data_type().name()
+            success, model, path = _Gtk.tree_get_row_drag_data(data)
+            if success:
+                assert model is self
+                iter = model.get_iter(path)
+                # Only allow ingredients to be dropped onto an ingredient group
+                return not isinstance(self.get_value(iter, 4), _IngredientGroup)
+        return False
+
+
+class _TreeViewSelectionFixer:
+    # Let's try to stop kittens from crying and fix some of GTK's inexplicable
+    # tree view selection behaviors
+
+    def __init__(self, tree, groups):
+        assert isinstance(tree, _Gtk.TreeView)
+
+        # Give keyboard focus to newly-selected rows
+        handler = tree.get_selection().connect('changed', self._focus_on_selected_row)
+
+        # Select the dropped row after drag-and-drop
+        tree.connect('drag-begin', self._start_drag)
+        tree.connect('drag-end', self._end_drag)
+
+        self._tree = tree
+        self._groups = groups
+        self._drag_destination = None
+        self._selection_handler = handler
+        self._insertion_handler = None
+
+
+    # For some reason, GTK doesn't give a newly-selected row the keyboard
+    # focus, so even though a user may click on a row with their mouse, the
+    # keyboard will continue to operate on whatever widget the user was
+    # previously interacting with. So let's fix this.
+    def _focus_on_selected_row(self, selection):
+        model, selected = selection.get_selected()
+        if selected:
+            path = model.get_path(selected)
+            tree = selection.get_tree_view()
+            tree.set_cursor(path, None, None)
+            tree.grab_focus()
+
+
+    # GTK has decided that, upon conclusion of a drag-and-drop operation, the
+    # current row should become whatever row is occupying the location you just
+    # dragged from instead of staying on the dropped row, which is visually
+    # quite confusing: it makes it difficult to determine where the row was
+    # dropped and even whether or not the operation succeeded. So we have to
+    # take care of this ourselves. Of course, GTK doesn't make it easy.
```

Comment:
```suggestion
    # take care of this ourselves.
```

I also have issues with Gtk. However, the code should remain neutral